prov/mrail/src/mrail_rma.c

```diff
@@ -0,0 +1,428 @@
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "mrail.h"
+
+static inline void mrail_iov_to_rail(const struct iovec *iov, const void *desc,
+		const uint32_t rail, struct iovec *out_iov, void **out_desc)
+{
+	const struct mrail_mr *mrail_mr = desc;
+
+	//TODO: add base address from mrail_mr
+	out_iov->iov_base 	= (void*)((uintptr_t)iov->iov_base);
+	out_iov->iov_len	= iov->iov_len;
+
+	*out_desc = (mrail_mr ? fi_mr_desc(mrail_mr->rails[rail].mr) : NULL);
+}
+
+static inline void mrail_rma_iov_to_rail(const struct fi_rma_iov *rma_iov,
+		const uint32_t rail, struct fi_rma_iov *out_rma_iov)
+{
+	struct mrail_addr_key *mr_map = (struct mrail_addr_key *)rma_iov->key;
+
+	//TODO: add base address from mrail_addr_key
+	out_rma_iov->addr 	= rma_iov->addr;
+	out_rma_iov->len	= rma_iov->len;
+	out_rma_iov->key	= mr_map[rail].key;
+}
+
+static void mrail_subreq_to_rail(struct mrail_subreq *subreq, uint32_t rail,
+		struct iovec *out_iovs, void **out_descs,
+		struct fi_rma_iov *out_rma_iovs)
+{
+	size_t i;
+
+	for (i = 0; i < subreq->iov_count; ++i) {
+		mrail_iov_to_rail(&subreq->iov[i], subreq->descs[i], rail,
+				&out_iovs[i], &out_descs[i]);
+	}
+
+	for (i = 0; i < subreq->rma_iov_count; ++i) {
+		mrail_rma_iov_to_rail(subreq->rma_iov, rail, &out_rma_iovs[i]);
+	}
+}
+
+static ssize_t mrail_try_posting_subreq(uint32_t rail,
+		struct mrail_subreq *subreq)
+{
+	ssize_t ret;
+	struct iovec rail_iov[MRAIL_IOV_LIMIT];
+	void *rail_descs[MRAIL_IOV_LIMIT];
+	struct fi_rma_iov rail_rma_iov[MRAIL_IOV_LIMIT];
+	struct fi_msg_rma msg;
+
+	struct mrail_req *req = subreq->parent;
+	struct mrail_ep *mrail_ep = req->mrail_ep;
+
+	mrail_subreq_to_rail(subreq, rail, rail_iov, rail_descs, rail_rma_iov);
+
+	msg.msg_iov		= rail_iov;
+	msg.desc		= rail_descs;
+	msg.iov_count		= subreq->iov_count;
+	msg.addr		= req->remote_addrs[rail];
+	msg.rma_iov		= rail_rma_iov;
+	msg.rma_iov_count	= subreq->rma_iov_count;
+	msg.context		= &subreq->context;
+	//TODO: we may not want to send immediate data with all subreqs
+	msg.data		= req->data;
+
+	switch (req->op_type) {
+	case FI_READ:
+		ret = fi_readmsg(mrail_ep->rails[rail].ep, &msg, req->flags);
+		break;
+	case FI_WRITE:
+		ret = fi_writemsg(mrail_ep->rails[rail].ep, &msg, req->flags);
+		break;
+	default:
+		FI_WARN(&mrail_prov, FI_LOG_EP_DATA,
+			"Unsupported operation: %d\n",
+			req->op_type);
+		ret = -FI_EINVAL;
+	};
+
+	return ret;
+}
+
+static ssize_t mrail_try_posting_req(struct mrail_req *req)
+{
+	size_t i;
+	uint32_t rail;
+	ssize_t ret = 0;
+
+	while (req->pending_subreq >= 0) {
+		/* Try all rails before giving up */
+		for (i = 0; i < req->mrail_ep->num_eps; ++i) {
+			rail = mrail_get_tx_rail(req->mrail_ep);
+
+			ret = mrail_try_posting_subreq(rail,
+					&req->subreqs[req->pending_subreq]);
+			if (ret != -FI_EAGAIN) {
+				break;
+			} else {
+				/* One of the rails is busy. Try progressing. */
+				mrail_poll_cq(req->mrail_ep->util_ep.tx_cq);
```

Comment:
This checks the cq even if the post succeeds.  I think some sort of restructuring is needed here, but it's not trivial enough for me to come up with it on the spot.

Reply:
Does it? If ret == 0, then we break the loop, no?

Reply:
oh, yeah :)